contracts/eden/src/auctions.cpp

```diff
@@ -0,0 +1,180 @@
+         double collection_fee;
+
+         uint64_t primary_key() const { return auction_id; };
+
+         eosio::checksum256 by_assets() const;
+      };
+      EOSIO_REFLECT(auction,
+                    auction_id,
+                    seller,
+                    asset_ids,
+                    end_time,
+                    assets_transferred,
+                    current_bid,
+                    current_bidder,
+                    claimed_by_seller,
+                    claimed_by_buyer,
+                    maker_marketplace,
+                    taker_marketplace,
+                    collection_name,
+                    collection_fee);
+
+      using auction_table_type = eosio::multi_index<
+          "auctions"_n,
+          auction,
+          eosio::indexed_by<
+              "byassets"_n,
+              eosio::const_mem_fun<auction, eosio::checksum256, &auction::by_assets>>>;
+   }  // namespace atomicmarket
+
+   void auctions::add_auction(uint64_t asset_id)
+   {
+      auction_tb.emplace(contract, [&](auto& row) {
+         row.asset_id() = asset_id;
+         row.last_known_end_time() =
+             eosio::current_time_point() + eosio::seconds(globals.get().auction_duration);
+      });
+   }
+
+   void auctions::add_auction(uint64_t asset_id, uint32_t end_time)
+   {
+      auction_tb.emplace(contract, [&](auto& row) {
+         row.asset_id() = asset_id;
+         row.last_known_end_time() = eosio::time_point_sec{end_time};
+      });
+   }
+
+   bool auctions::has_auction(uint64_t asset_id)
+   {
+      return auction_tb.find(asset_id) != auction_tb.end();
+   }
+
+   uint32_t auctions::finish_auctions(uint32_t max_steps)
+   {
+      atomicmarket::auction_table_type market_tb(atomic_market_account,
+                                                 atomic_market_account.value);
+      const auto& asset_id_idx = market_tb.get_index<"byassets"_n>();
+
+      const auto& end_time_idx = auction_tb.get_index<"bytime"_n>();
+      auto iter = end_time_idx.begin();
+      auto end = end_time_idx.end();
+      eosio::time_point_sec current_time = eosio::current_time_point();
+      for (; max_steps > 0 && iter != end && iter->last_known_end_time() < current_time;
+           --max_steps)
+      {
+         const auto& auction = *iter++;
+         // Find the auction in atomicmarket if it exists.  There may be multiple auction rows
+         // with these assets, as the asset can be reauctioned before we claim the funds.
+         // We can distinguish the right auction by seller.
+         auto key = eosio::sha256(reinterpret_cast<const char*>(&auction.asset_id()),
+                                  sizeof(auction.asset_id()));
+         auto market_iter = asset_id_idx.lower_bound(key);
+         while (true)
+         {
+            if (market_iter->asset_ids.size() != 1 ||
+                market_iter->asset_ids.front() != auction.asset_id())
+            {
+               market_iter = asset_id_idx.end();
```

Comment:
This assumes there won't be a hash collision. Probably ok because:
* It will be astronomically rare
* If it happens, the result isn't a big deal

Reply:
Also, any hash collision means that the security of sha256 is compromised, and we have much bigger problems.

Reply:
One of the eosio chains, I forgot which, had a collision between 2 different transactions. They were far enough apart that nodeos let them through. Several history solutions fell over and had to get workarounds.

Reply:
Were they deferred transactions?

Reply:
Identical transactions will have the same hash, and deferred transactions have tapos cleared.

Reply:
I don't remember